includes/class-newspack-segments-model.php

```diff
@@ -0,0 +1,616 @@
+		}
+
+		$popups = Newspack_Popups_Model::retrieve_popups( true, true );
+		foreach ( $popups as $popup ) {
+			$meta_value = get_post_meta( $popup['id'], 'selected_segment_id', true );
+			if ( $meta_value ) {
+				// Create a backup of the old value.
+				update_post_meta( $popup['id'], 'selected_segment_id_bkp', $meta_value );
+				foreach ( $id_mapping as $old_id => $new_id ) {
+					$meta_value = str_replace( $old_id, $new_id, $meta_value );
+				}
+				update_post_meta( $popup['id'], 'selected_segment_id', $meta_value );
+			}
+		}
+	}
+
+	/**
+	 * Register the segments taxonomy
+	 *
+	 * @return void
+	 */
+	public static function register_segments_taxonomy() {
+		$labels = array(
+			'name'              => _x( 'Segments', 'taxonomy general name', 'newspack-popups' ),
+			'singular_name'     => _x( 'Segment', 'taxonomy singular name', 'newspack-popups' ),
+			'search_items'      => __( 'Search Segments', 'newspack-popups' ),
+			'all_items'         => __( 'All Segments', 'newspack-popups' ),
+			'parent_item'       => __( 'Parent Segment', 'newspack-popups' ),
+			'parent_item_colon' => __( 'Parent Segment:', 'newspack-popups' ),
+			'edit_item'         => __( 'Edit Segment', 'newspack-popups' ),
+			'update_item'       => __( 'Update Segment', 'newspack-popups' ),
+			'add_new_item'      => __( 'Add New Segment', 'newspack-popups' ),
+			'new_item_name'     => __( 'New Segment Name', 'newspack-popups' ),
+			'menu_name'         => __( 'Segments', 'newspack-popups' ),
+		);
+
+		$args = array(
+			'labels'             => $labels,
+			'description'        => __( 'Segments for popups', 'newspack-popups' ),
+			'hierarchical'       => false,
+			'public'             => false,
+			'show_ui'            => false,
+			'show_in_menu'       => false,
+			'show_in_nav_menus'  => false,
+			'show_tagcloud'      => false,
+			'show_in_quick_edit' => false,
+			'show_admin_column'  => false,
+			'rewrite'            => false,
+		);
+
+		register_taxonomy( self::TAX_SLUG, array( Newspack_Popups::NEWSPACK_POPUPS_CPT ), $args );
+
+		self::register_meta_fields();
+	}
+
+	/**
+	 * Gets the Schema for the meta fields
+	 *
+	 * @return array The schema.
+	 */
+	public static function get_meta_schema() {
+		return [
+			'priority'      => [
+				'name'     => 'priority',
+				'type'     => 'integer',
+				'required' => false,
+				'default'  => PHP_INT_MAX,
+			],
+			'created_at'    => [
+				'name'      => 'created_at',
+				'type'      => 'string',
+				'required'  => false,
+				'maxLength' => 10,
+				'pattern'   => '^\d{4}-\d\d-\d\d$',
+			],
+			'updated_at'    => [
+				'name'      => 'created_at',
+				'type'      => 'string',
+				'required'  => false,
+				'maxLength' => 10,
+				'pattern'   => '^\d{4}-\d\d-\d\d$',
+			],
+			'configuration' => [
+				'name'                 => 'configuration',
+				'type'                 => 'object',
+				'required'             => true,
+				'additionalProperties' => false,
+				'properties'           => [
```

Comment:
Moving to the non-hardcoded criteria approach this could be tweaked to maybe support an array of criteria configurations instead of a strict definition.

Reply:
yes!